### ISSUE TITLE
[promise-function-async] Should ignore cases when is running inside .then

### ISSUE THREAD

**Issue by tinovyatkin** [typescript-eslint/typescript-eslint#2383]
I believe cases when a function is returning Promise inside of context of `.then` should be ignored:

```ts
replSet
  .waitUntilRunning()
  .then(() => replSet.getConnectionString()) // this function is returning a promise but inside promise chain
  .then(connectionString =>
    console.info(
      "MongoDB Replica Set Started Successfully: %s",
      connectionString,
    ),
  )
  .catch(err => console.error(err));
```

**Comment by bradzacher** [comment]
How come you think this function shouldn't be marked as async?
It's still an async function even if it's being passed to `then`.

`.then` is just a function which accepts a value or a promise, so there's nothing special about it.

---

As an aside - why not just leverage `async`/`await` here for flatter code?

```ts
try {
  await replSet.waitUntilRunning();
  const connectionString = await replSet.getConnectionString();
  console.info(
      "MongoDB Replica Set Started Successfully: %s",
      connectionString,
  );
} catch (err) {
  console.error(err);
}
```

**Comment by tinovyatkin** [comment]
@bradzacher 
I know that it can be refactored into `async/await`, but due to lack of top-level await or `@typescript-eslint/no-floating-promises` let's assume we would like to keep the promise chain. So, what this rule recommends in this case is following, definitely bad practice code that will be conflicting with `no-async-promise-executor` from `eslint:recommended` :

```ts
replSet
  .waitUntilRunning()
  .then(async () => await replSet.getConnectionString()) // this function is returning a promise but inside promise chain
  .then(connectionString =>
    console.info(
      "MongoDB Replica Set Started Successfully: %s",
      connectionString,
    ),
  )
  .catch(err => console.error(err));
```

**Comment by bradzacher** [comment]
> definitely bad practice code that will be conflicting with no-async-promise-executor

[I think you have misunderstood the `no-async-promise-executor` rule.](https://eslint.org/demo#eyJ0ZXh0IjoiLyogZXNsaW50IG5vLWFzeW5jLXByb21pc2UtZXhlY3V0b3I6ICdlcnJvcicgKi9cblxucmVwbFNldFxuICAud2FpdFVudGlsUnVubmluZygpXG4gIC50aGVuKGFzeW5jICgpID0+IGF3YWl0IHJlcGxTZXQuZ2V0Q29ubmVjdGlvblN0cmluZygpKSAvLyBubyB3YXJuaW5nIGZyb20gdGhlIHJ1bGUsIGJ1dCBmcm9tIG5vLWFzeW5jLXByb21pc2UtZXhlY3V0b3Igbm93XG4gIC50aGVuKGNvbm5lY3Rpb25TdHJpbmcgPT5cbiAgICBjb25zb2xlLmluZm8oXG4gICAgICBcIk1vbmdvREIgUmVwbGljYSBTZXQgU3RhcnRlZCBTdWNjZXNzZnVsbHk6ICVzXCIsXG4gICAgICBjb25uZWN0aW9uU3RyaW5nLFxuICAgICksXG4gIClcbiAgLmNhdGNoKGVyciA9PiBjb25zb2xlLmVycm9yKGVycikpO1xuXG4iLCJvcHRpb25zIjp7InBhcnNlck9wdGlvbnMiOnsiZWNtYVZlcnNpb24iOjksInNvdXJjZVR5cGUiOiJtb2R1bGUiLCJlY21hRmVhdHVyZXMiOnsianN4Ijp0cnVlfX0sInJ1bGVzIjp7fSwiZW52Ijp7fX19)
That rule is about ensuring you don't pass async functions into the promise constructor, not the `then`.

Passing an `async` function into `then` is not bad practice - it's standard practice.

---

> but due to lack of top-level await or @typescript-eslint/no-floating-promises

The following code is valid for the `no-floating-promises` rule, because [`ignoreVoid`](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/no-floating-promises.md#ignorevoid) is on by default

```ts
async function foo() {
  try {
    await replSet.waitUntilRunning();
    const connectionString = await replSet.getConnectionString();
    console.info(
      "MongoDB Replica Set Started Successfully: %s",
      connectionString,
    );
  } catch (err) {
    console.error(err);
  }
}

void foo();
```
